include/soft_limiter.h

```diff
@@ -0,0 +1,255 @@
+
+Use:
+
+The SoftLimiter reads and writes arrays of the same length, which is
+defined during object creation as a template size. For example:
+
+  SoftLimiter<48> limiter;
+
+You can then repeatedly call:
+  limiter.Apply(in_samples, out_samples);
+
+Where in_samples is a std::array<float, 48> and out_samples is a 
+std::array<int16_t, 48>. The limiter will either copy or limit the in_samples
+into the out_samples array.
+
+The PrintStats function will make mixer suggestions if the in_samples
+were (at most) 40% under the allowed bounds, in which case the recommendation
+will describe how to scale up the channel amplitude. 
+
+On the other hand, if the limiter found itself scaling down more than 10% of
+the inbound stream, then it will in turn recommend how to scale down the
+channel. 
+*/
+
+#include "dosbox.h"
+#include "logging.h"
+#include "mixer.h"
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <limits>
+#include <cmath>
+#include <string>
+
+template <size_t array_frames>
+class SoftLimiter {
+public:
+	SoftLimiter() = delete;
+	SoftLimiter(const std::string &name, const AudioFrame &scale);
+
+	using in_array_t = std::array<float, array_frames * 2>; // 2 for stereo
+	using out_array_t = std::array<int16_t, array_frames * 2>; // 2 for stereo
+
+	void Apply(const in_array_t &in, out_array_t &out, uint16_t frames) noexcept;
+	const AudioFrame &GetPeaks() const noexcept { return peak; }
+	void PrintStats(const AudioFrame &external_scale) const;
+	void Reset() noexcept;
+
+private:
+	// Amplitude level constants
+	using out_limits = std::numeric_limits<int16_t>;
+	constexpr static float upper_bound = static_cast<float>(out_limits::max());
+	constexpr static float lower_bound = static_cast<float>(out_limits::min());
+
+	bool FindPeaks(const in_array_t &stream, uint16_t frames) noexcept;
+	void Release() noexcept;
+
+	std::string channel_name = {};
+	AudioFrame limit_scale = {1, 1}; // real-time limit applied to stream
+	AudioFrame peak = {1, 1};        // holds real-time peak amplitudes
+	const AudioFrame &prescale;      // scale before operating on the stream
+	int limited_ms = 0;              // milliseconds that needed limiting
+	int non_limited_ms = 0; // milliseconds that didn't need limiting
+};
+
+template <size_t array_frames>
+SoftLimiter<array_frames>::SoftLimiter(const std::string &name, const AudioFrame &scale)
+        : channel_name(name),
+          prescale(scale)
+{
+	static_assert(array_frames > 0, "need some quantity of frames");
+	static_assert(array_frames < 16384, "too many frames adds audible latency");
+}
+
+template <size_t array_frames>
+bool SoftLimiter<array_frames>::FindPeaks(const in_array_t &stream,
+                                          const uint16_t samples) noexcept
+{
+	auto val = stream.begin();
+	const auto val_end = stream.begin() + samples;
+	while (val < val_end) {
+		// Left channel
+		const auto amplitude_left = fabsf(*val) * prescale.left;
```

Comment:
Optimization: don't multiple by prescale inside the loop and don't compare peak.* Inside the loop. Instead, get the local max value for the left and right channels, then compare those single values (multiplied by prescale) with peak.* As a one-time comparison outside the loop.